Retry after encountering a checksum failure in GCSFileSystem.cat?

We've been using gcsfs fairly heavily for loading data from zarr stores in the cloud; in general it has worked really well.  A basic pattern might look something like this:

```python
import gcsfs
import xarray

PROJECT = "example-project"
STORE = "gs://path-to-store/store.zarr"

gcs = gcsfs.GCSFileSystem(project=PROJECT)
mapper = gcs.get_mapper(STORE)
ds = xarray.open_zarr(mapper)
result = ds.mean().compute()
```

Every now and then -- I honestly don't know what dictates it -- there are days when we encounter transient errors like this when performing the computation:

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)

... (very long traceback)

~/miniconda3/envs/fv3net/lib/python3.7/site-packages/gcsfs/core.py in cat(self, path)
    888                 if key == "md5":
    889                     md = b64decode(val)
--> 890                     assert md5(r.content).digest() == md, "Checksum failure"
    891         return r.content
    892 
AssertionError: Checksum failure
```

Typically one can retry the script a few times and eventually the full computation succeeds.  This can be a bit frustrating, however, because particularly for long-running computations, many downloads will be performed before one of these errors is hit, and re-running the script means starting over from scratch.  Would it be appropriate to add some logic to `GCSFileSystem.cat` that would retry a download if a checksum failure occurred?  I believe there are other places in gcsfs where retry logic is used to work around these kinds of transient issues.

Apologies for not being able to provide a minimal working example -- the transient nature of the issue makes that difficult.  

cc: @nbren12 @oliverwm1 @brianhenn

Are you able to determine if, for these cases, the downloaded data really is wrong, perhaps truncated? It would be interesting to see if the http response headers claim partial data.

Not opposed to retry on checksum failure.

On July 10, 2020 6:27:02 PM EDT, Spencer Clark <notifications@github.com> wrote:
>We've been using gcsfs fairly heavily for loading data from zarr stores
>in the cloud; in general it has worked really well.  A basic pattern
>might look something like this:
>
>```python
>import gcsfs
>import xarray
>
>PROJECT = "example-project"
>STORE = "gs://path-to-store/store.zarr"
>
>gcs = gcsfs.GCSFileSystem(project=PROJECT)
>mapper = gcs.get_mapper(STORE)
>ds = xarray.open_zarr(mapper)
>result = ds.mean().compute()
>```
>
>Every now and then -- I honestly don't know what dictates it -- there
>are days when we encounter transient errors like this when performing
>the computation:
>
>```
>---------------------------------------------------------------------------
>AssertionError                            Traceback (most recent call
>last)
>
>... (very long traceback)
>
>~/miniconda3/envs/fv3net/lib/python3.7/site-packages/gcsfs/core.py in
>cat(self, path)
>    888                 if key == "md5":
>    889                     md = b64decode(val)
>--> 890                     assert md5(r.content).digest() == md,
>"Checksum failure"
>    891         return r.content
>    892 
>AssertionError: Checksum failure
>```
>
>Typically one can retry the script a few times and eventually the full
>computation succeeds.  This can be a bit frustrating, however, because
>particularly for long-running computations, many downloads will be
>performed before one of these errors is hit, and re-running the script
>means starting over from scratch.  Would it be appropriate to add some
>logic to `GCSFileSystem.cat` that would retry a download if a checksum
>failure occurred?  I believe there are other places in gcsfs where
>retry logic is used to work around these kinds of transient issues.
>
>Apologies for not being able to provide a minimal working example --
>the transient nature of the issue makes that difficult.  
>
>cc: @nbren12 @oliverwm1 @brianhenn
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/dask/gcsfs/issues/267

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

Also +1 on retry by default

On Fri, Jul 10, 2020 at 5:32 PM Martin Durant <notifications@github.com>
wrote:

> Are you able to determine if, for these cases, the downloaded data really
> is wrong, perhaps truncated? It would be interesting to see if the http
> response headers claim partial data.
>
> Not opposed to retry on checksum failure.
>
> On July 10, 2020 6:27:02 PM EDT, Spencer Clark <notifications@github.com>
> wrote:
> >We've been using gcsfs fairly heavily for loading data from zarr stores
> >in the cloud; in general it has worked really well. A basic pattern
> >might look something like this:
> >
> >```python
> >import gcsfs
> >import xarray
> >
> >PROJECT = "example-project"
> >STORE = "gs://path-to-store/store.zarr"
> >
> >gcs = gcsfs.GCSFileSystem(project=PROJECT)
> >mapper = gcs.get_mapper(STORE)
> >ds = xarray.open_zarr(mapper)
> >result = ds.mean().compute()
> >```
> >
> >Every now and then -- I honestly don't know what dictates it -- there
> >are days when we encounter transient errors like this when performing
> >the computation:
> >
> >```
>
> >---------------------------------------------------------------------------
> >AssertionError Traceback (most recent call
> >last)
> >
> >... (very long traceback)
> >
> >~/miniconda3/envs/fv3net/lib/python3.7/site-packages/gcsfs/core.py in
> >cat(self, path)
> > 888 if key == "md5":
> > 889 md = b64decode(val)
> >--> 890 assert md5(r.content).digest() == md,
> >"Checksum failure"
> > 891 return r.content
> > 892
> >AssertionError: Checksum failure
> >```
> >
> >Typically one can retry the script a few times and eventually the full
> >computation succeeds. This can be a bit frustrating, however, because
> >particularly for long-running computations, many downloads will be
> >performed before one of these errors is hit, and re-running the script
> >means starting over from scratch. Would it be appropriate to add some
> >logic to `GCSFileSystem.cat` that would retry a download if a checksum
> >failure occurred? I believe there are other places in gcsfs where
> >retry logic is used to work around these kinds of transient issues.
> >
> >Apologies for not being able to provide a minimal working example --
> >the transient nature of the issue makes that difficult.
> >
> >cc: @nbren12 @oliverwm1 @brianhenn
> >
> >--
> >You are receiving this because you are subscribed to this thread.
> >Reply to this email directly or view it on GitHub:
> >https://github.com/dask/gcsfs/issues/267
>
> --
> Sent from my Android device with K-9 Mail. Please excuse my brevity.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/gcsfs/issues/267#issuecomment-656949829>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACKZTDBKXEIPD6SPYRFA7LR26XIBANCNFSM4OW7J4IA>
> .
>

--- a/gcsfs/core.py
+++ b/gcsfs/core.py
@@ -1181,9 +1181,6 @@ class GCSFileSystem(AsyncFileSystem):
         r: requests response object
         path: associated URL path, for error messages
         """
-        if headers is None:
-            headers = {}
-
         if status >= 400:
             error = None
             try:
@@ -1208,7 +1205,7 @@ class GCSFileSystem(AsyncFileSystem):
             else:
                 raise RuntimeError(msg)
         else:
-            if "X-Goog-Hash" in headers:
+            if headers is not None and "X-Goog-Hash" in headers:
                 # if header includes md5 hash, check that data matches
                 bits = headers["X-Goog-Hash"].split(",")
                 for bit in bits:

--- a/gcsfs/core.py
+++ b/gcsfs/core.py
@@ -1172,7 +1172,7 @@ class GCSFileSystem(AsyncFileSystem):
         else:
             return path.split("/", 1)
 
-    def validate_response(self, status, content, json, path, headers):
+    def validate_response(self, status, content, json, path, headers=None):
         """
         Check the requests object r, raise error if it's not ok.
 
@@ -1181,6 +1181,9 @@ class GCSFileSystem(AsyncFileSystem):
         r: requests response object
         path: associated URL path, for error messages
         """
+        if headers is None:
+            headers = {}
+
         if status >= 400:
             error = None
             try:

--- a/gcsfs/core.py
+++ b/gcsfs/core.py
@@ -24,7 +24,6 @@ import posixpath
 import requests
 import pickle
 import re
-import time
 import requests
 import threading
 import warnings

--- a/gcsfs/core.py
+++ b/gcsfs/core.py
@@ -34,7 +34,7 @@ import weakref
 from requests.exceptions import RequestException, ProxyError
 from fsspec.asyn import sync_wrapper, sync, AsyncFileSystem
 from fsspec.implementations.http import get_client
-from .utils import HttpError, is_retriable
+from .utils import ChecksumError, HttpError, is_retriable
 from . import __version__ as version
 
 logger = logging.getLogger(__name__)
@@ -496,9 +496,9 @@ class GCSFileSystem(AsyncFileSystem):
                     except Exception:
                         json = None
 
-                self.validate_response(status, contents, json, path)
+                self.validate_response(status, contents, json, path, headers)
                 break
-            except (HttpError, RequestException, GoogleAuthError) as e:
+            except (HttpError, RequestException, GoogleAuthError, ChecksumError) as e:
                 json = None
                 if (
                     isinstance(e, HttpError)
@@ -816,14 +816,6 @@ class GCSFileSystem(AsyncFileSystem):
         """ Simple one-shot get of file data """
         u2 = self.url(path)
         headers, out = await self._call("GET", u2)
-        if "X-Goog-Hash" in headers:
-            # if header includes md5 hash, check that data matches
-            bits = headers["X-Goog-Hash"].split(",")
-            for bit in bits:
-                key, val = bit.split("=", 1)
-                if key == "md5":
-                    md = b64decode(val)
-                    assert md5(out).digest() == md, "Checksum failure"
         return out
 
     def getxattr(self, path, attr):
@@ -1181,7 +1173,7 @@ class GCSFileSystem(AsyncFileSystem):
         else:
             return path.split("/", 1)
 
-    def validate_response(self, status, content, json, path):
+    def validate_response(self, status, content, json, path, headers):
         """
         Check the requests object r, raise error if it's not ok.
 
@@ -1213,6 +1205,16 @@ class GCSFileSystem(AsyncFileSystem):
                 raise HttpError({"code": status})
             else:
                 raise RuntimeError(msg)
+        else:
+            if "X-Goog-Hash" in headers:
+                # if header includes md5 hash, check that data matches
+                bits = headers["X-Goog-Hash"].split(",")
+                for bit in bits:
+                    key, val = bit.split("=", 1)
+                    if key == "md5":
+                        md = b64decode(val)
+                        if md5(content).digest() != md:
+                            raise ChecksumError("Checksum failure")
 
 
 GCSFileSystem.load_tokens()

--- a/gcsfs/tests/test_core.py
+++ b/gcsfs/tests/test_core.py
@@ -1,5 +1,7 @@
 # -*- coding: utf-8 -*-
 
+import base64
+import hashlib
 import io
 from builtins import FileNotFoundError
 from itertools import chain
@@ -11,7 +13,7 @@ import requests
 from fsspec.utils import seek_delimiter
 from requests.exceptions import ProxyError
 
-from gcsfs.utils import HttpError
+from gcsfs.utils import ChecksumError, HttpError
 
 from gcsfs.tests.settings import (
     TEST_PROJECT,
@@ -859,30 +861,37 @@ def test_raise_on_project_mismatch(mock_auth):
 
 def test_validate_response():
     gcs = GCSFileSystem(token="anon")
-    gcs.validate_response(200, None, None, "/path")
+    gcs.validate_response(200, None, None, "/path", {})
 
     # HttpError with no JSON body
     with pytest.raises(HttpError) as e:
-        gcs.validate_response(503, b"", None, "/path")
+        gcs.validate_response(503, b"", None, "/path", {})
     assert e.value.code == 503
     assert e.value.message == ""
 
     # HttpError with JSON body
     j = {"error": {"code": 503, "message": b"Service Unavailable"}}
     with pytest.raises(HttpError) as e:
-        gcs.validate_response(503, None, j, "/path")
+        gcs.validate_response(503, None, j, "/path", {})
     assert e.value.code == 503
     assert e.value.message == b"Service Unavailable"
 
     # 403
     j = {"error": {"message": "Not ok"}}
     with pytest.raises(IOError, match="Forbidden: /path\nNot ok"):
-        gcs.validate_response(403, None, j, "/path")
+        gcs.validate_response(403, None, j, "/path", {})
 
     # 404
     with pytest.raises(FileNotFoundError):
-        gcs.validate_response(404, b"", None, "/path")
+        gcs.validate_response(404, b"", None, "/path", {})
 
     # 502
     with pytest.raises(ProxyError):
-        gcs.validate_response(502, b"", None, "/path")
+        gcs.validate_response(502, b"", None, "/path", {})
+
+    # ChecksumError
+    md5 = repr(base64.b64encode(hashlib.md5(b"foo").digest()))[2:-1]
+    with pytest.raises(ChecksumError):
+        gcs.validate_response(
+            0, b"f", None, "/path", {"X-Goog-Hash": f"md5={md5}"}
+        )

--- a/gcsfs/tests/test_core.py
+++ b/gcsfs/tests/test_core.py
@@ -861,33 +861,33 @@ def test_raise_on_project_mismatch(mock_auth):
 
 def test_validate_response():
     gcs = GCSFileSystem(token="anon")
-    gcs.validate_response(200, None, None, "/path", {})
+    gcs.validate_response(200, None, None, "/path")
 
     # HttpError with no JSON body
     with pytest.raises(HttpError) as e:
-        gcs.validate_response(503, b"", None, "/path", {})
+        gcs.validate_response(503, b"", None, "/path")
     assert e.value.code == 503
     assert e.value.message == ""
 
     # HttpError with JSON body
     j = {"error": {"code": 503, "message": b"Service Unavailable"}}
     with pytest.raises(HttpError) as e:
-        gcs.validate_response(503, None, j, "/path", {})
+        gcs.validate_response(503, None, j, "/path")
     assert e.value.code == 503
     assert e.value.message == b"Service Unavailable"
 
     # 403
     j = {"error": {"message": "Not ok"}}
     with pytest.raises(IOError, match="Forbidden: /path\nNot ok"):
-        gcs.validate_response(403, None, j, "/path", {})
+        gcs.validate_response(403, None, j, "/path")
 
     # 404
     with pytest.raises(FileNotFoundError):
-        gcs.validate_response(404, b"", None, "/path", {})
+        gcs.validate_response(404, b"", None, "/path")
 
     # 502
     with pytest.raises(ProxyError):
-        gcs.validate_response(502, b"", None, "/path", {})
+        gcs.validate_response(502, b"", None, "/path")
 
     # ChecksumError
     md5 = repr(base64.b64encode(hashlib.md5(b"foo").digest()))[2:-1]

--- a/gcsfs/tests/test_core.py
+++ b/gcsfs/tests/test_core.py
@@ -892,6 +892,4 @@ def test_validate_response():
     # ChecksumError
     md5 = repr(base64.b64encode(hashlib.md5(b"foo").digest()))[2:-1]
     with pytest.raises(ChecksumError):
-        gcs.validate_response(
-            0, b"f", None, "/path", {"X-Goog-Hash": f"md5={md5}"}
-        )
+        gcs.validate_response(0, b"f", None, "/path", {"X-Goog-Hash": f"md5={md5}"})

--- a/gcsfs/utils.py
+++ b/gcsfs/utils.py
@@ -18,6 +18,7 @@ class HttpError(Exception):
 
 class ChecksumError(Exception):
     """Raised when the md5 hash of the content does not match the header."""
+
     pass
 
 
@@ -30,7 +31,7 @@ RETRIABLE_EXCEPTIONS = (
     requests.exceptions.SSLError,
     requests.exceptions.ContentDecodingError,
     google.auth.exceptions.RefreshError,
-    ChecksumError
+    ChecksumError,
 )
 
 

--- a/gcsfs/utils.py
+++ b/gcsfs/utils.py
@@ -16,6 +16,11 @@ class HttpError(Exception):
         super(HttpError, self).__init__(self.message)
 
 
+class ChecksumError(Exception):
+    """Raised when the md5 hash of the content does not match the header."""
+    pass
+
+
 RETRIABLE_EXCEPTIONS = (
     requests.exceptions.ChunkedEncodingError,
     requests.exceptions.ConnectionError,
@@ -25,6 +30,7 @@ RETRIABLE_EXCEPTIONS = (
     requests.exceptions.SSLError,
     requests.exceptions.ContentDecodingError,
     google.auth.exceptions.RefreshError,
+    ChecksumError
 )
 
 

